js/bitopro.js

```diff
@@ -0,0 +1,1151 @@
+            'datetime': undefined,
+            'network': undefined,
+            'addressFrom': undefined,
+            'address': address,
+            'addressTo': address,
+            'tagFrom': undefined,
+            'tag': undefined,
+            'tagTo': undefined,
+            'type': undefined,
+            'amount': amount,
+            'currency': code,
+            'status': undefined,
+            'updated': undefined,
+            'fee': {
+                'currency': code,
+                'cost': fee,
+                'rate': undefined,
+            },
+        };
+    }
+
+    async withdraw (code, amount, address, tag = undefined, params = {}) {
+        [ tag, params ] = this.handleWithdrawTagAndParams (tag, params);
+        await this.loadMarkets ();
+        this.checkAddress (address);
+        const currency = this.currency (code);
+        const request = {
+            'currency': currency['id'],
+            'amount': this.numberToString (amount),
+            'address': address,
+            'protocol': '',
+        };
+        if (tag !== undefined) {
+            request['message'] = tag;
+        }
+        const response = await this.privatePostWalletWithdrawCurrency (this.extend (request, params));
+        const result = this.safeValue (response, 'data', {});
+        //
+        //     {
+        //         "data":{
+        //             "serial":"20220215BW14069838",
+        //             "currency":"USDT",
+        //             "protocol":"TRX",
+        //             "address":"TKrwMaZaGiAvtXCFT41xHuusNcs4LPWS7w",
+        //             "amount":"8",
+        //             "fee":"2",
+        //             "total":"10"
+        //         }
+        //     }
+        //
+        return this.parseTransaction (result, currency);
+    }
+
+    sign (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
+        let url = '/' + this.implodeParams (path, params);
+        const query = this.omit (params, this.extractParams (path));
+        if (headers === undefined) {
+            headers = {};
+        }
+        headers['X-BITOPRO-API'] = 'ccxt';
+        if (api === 'private') {
+            this.checkRequiredCredentials ();
+            if (method === 'POST') {
+                body = this.json (params);
+                const payload = this.stringToBase64 (body);
+                const signature = this.hmac (payload, this.secret, 'sha384', 'hex');
+                headers['X-BITOPRO-APIKEY'] = this.apiKey;
+                headers['X-BITOPRO-PAYLOAD'] = payload;
+                headers['X-BITOPRO-SIGNATURE'] = signature;
+            } else if (method === 'GET' || method === 'DELETE') {
+                if (Object.keys (query).length) {
+                    url += '?' + this.urlencode (query);
+                }
+                const nonce = this.milliseconds ();
+                let rawData = {
+                    'identity': this.email,
```

Comment:
Signing is failing in python because of this, since we can successfully authenticate (js and php) without an email, I'm not sure we should assume that an email variable is always defined

![image](https://user-images.githubusercontent.com/43336371/154232547-cf47c15c-4adb-48ff-9f48-7d775e0ee70d.png)